### PR TITLE
[MERGE] Add new scorer: MixValScorer

### PR DESCRIPTION
**Pull request by YanisLalou**
Implementation of the paper:
https://proceedings.neurips.cc/paper_files/paper/2023/file/7721f1fea280e9ffae528dc78c732576-Paper-Conference.pdf

### CODE REVIEW

File: skada/metrics.py [hunk 1]
```diff
@@ -619,3 +619,125 @@ def _score(self, estimator, X, y, sample_domain=None):
...
+    MixVal scorer for unsupervised domain adaptation.
+
+    This scorer uses mixup to create mixed samples from the target domain,
+    and evaluates the model's consistency on these mixed samples.
+
+    See [32]_ for details.
+
+    Parameters
+    ----------
+    lmbd : float, default=0.55
```
Comment (tgnassou) [inline skada/metrics.py:635]: I don't know if sklearn uses `lmbd` name. Maybe switch to alpha or mixup_param is better ?

File: skada/metrics.py [hunk 2]
```diff
@@ -619,3 +619,125 @@ def _score(self, estimator, X, y, sample_domain=None):
...
+        # Generate mixed samples
+        X_mixed, y_mixed = self._generate_mixed_samples(X[~source_idx], y_pred)
+
+        # Get predictions for mixed samples
+        y_pred_mixed = estimator.predict_proba(
+            X_mixed, sample_domain=np.full(X_mixed.shape[0], -1)
+        )
+
+        # Calculate ICE score
+        ice_score = accuracy_score(y_mixed.argmax(axis=1), y_pred_mixed.argmax(axis=1))
```
Comment (tgnassou) [inline skada/metrics.py:737]: In the paper, they don't compute accuracy on all the mixed samples. They separate samples with the same `y_pred` and a different one. Then, they compute the accuracy score for diff_mixed_samples and same_mixed_sample and take the mean of the two scores. See Appendix A of the paper.

File: skada/metrics.py [hunk 3]
```diff
@@ -619,3 +619,131 @@ def _score(self, estimator, X, y, sample_domain=None):
...
+        X_target = X[~source_idx]
+
+        rng = check_random_state(self.random_state)
+        rand_idx = rng.permutation(X_target.shape[0])
+
+        # Get predictions for target samples
+        pred_a = estimator.predict_proba(
+            X_target, sample_domain=sample_domain[~source_idx]
+        )
+        pl_a = pred_a.argmax(axis=1)
```
Comment (tgnassou) [inline skada/metrics.py:712]: I don't really like the name `pl_a` and `pl_b` to be consistent we can call them `labels_a` and `labels_b`?

File: skada/metrics.py [hunk 4]
```diff
@@ -619,3 +619,131 @@ def _score(self, estimator, X, y, sample_domain=None):
...
+        pl_b = pl_a[rand_idx]
+
+        # Intra-cluster and inter-cluster mixup
+        same_idx = (pl_a == pl_b).nonzero()[0]
+        diff_idx = (pl_a != pl_b).nonzero()[0]
+
+        # Mixup with images and hard pseudo labels
+        mix_inputs = self.alpha * X_target + (1 - self.alpha) * X_target[rand_idx]
+        mix_labels = pl_a if self.alpha > 0.5 else pl_b
```
Comment (tgnassou) [inline skada/metrics.py:721]: In the pseudo-code they are doing that, but neither in the paper nor in the GitHub they are using this formula for the mix_labels. They're using the formula with `alpha`, maybe it is the same ?

Comment (YanisLalou) [reply skada/metrics.py:721]: Yes I agree, in the paper they're using the same formula as for X. However before going forward with the code, we need to agree what we should follow for all the implementation: The paper, the pseudo-code or the git code ?

**Pull request merged**